code/modules/ai/ai_node_pathfinder.dm

```diff
@@ -0,0 +1,68 @@
+GLOBAL_LIST_EMPTY(goal_nodes)
+
+/datum/node_path
+	var/distance_to_goal
+	var/distance_walked
+	var/obj/effect/ai_node/current_node
+	var/obj/effect/ai_node/previous_node
+
+/datum/node_path/New(obj/effect/ai_node/previous_node, obj/effect/ai_node/current_node, obj/effect/ai_node/goal_node, old_distance_walked)
+	. = ..()
+	distance_to_goal = get_dist_euclide_square(current_node, goal_node)
+	distance_walked = old_distance_walked + get_dist_euclide_square(current_node, previous_node)
+	src.current_node = current_node
+	src.previous_node = previous_node
+
+///Returns the most optimal nodes path to get from starting node to goal node
+/proc/get_node_path(obj/effect/ai_node/starting_node, obj/effect/ai_node/goal_node)
+	if(starting_node.z != goal_node.z)
+		CRASH("Start node and goal node were not on the same z level")
+	if(starting_node == goal_node)
+		CRASH("Start node and goal node were identical")
+	var/list/datum/node_path/paths_to_check = list()
+	var/obj/effect/ai_node/current_node = starting_node
+	var/list/datum/node_path/paths_checked = list()
+	var/datum/node_path/current_path
+	while(current_node != goal_node)
+		for(var/obj/effect/ai_node/node_to_check AS in current_node.adjacent_nodes)
+			if(paths_to_check[node_to_check] || paths_checked[node_to_check]) //We already found a better path to get to this node
+				continue
+			paths_to_check[node_to_check] = new/datum/node_path(current_node, node_to_check, goal_node, current_path?.distance_walked)
```

Comment:
whats going on here does this create looping loops for finding the target path

Reply:
see wiki link